### PR TITLE
Add `inputs.cache-save-always`

### PR DESCRIPTION
**Pull request by stackptr**
Adds an input that borrows the `save-always` input from [v4 of `actions/cache`][0], which will run the post step to save the cache even if a failure occurs. Using the added `cache-save-always` input saves the cache when a build fails, which will speed up subsequent runs which restore the partial build. This comes at the cost of requiring a build when there is an exact `key` match, since the cache from a failed build will have the same `key` as a completed build.

[0]: https://github.com/actions/cache/tree/main#v4

### CODE REVIEW

**Comment by stackptr** [comment]
This will require a major version since this changes the noop-behavior on a full cache hit. If the trade-offs seem worthwhile (and I'm not missing anything about how Stack might treat these artifacts) then I'll bump the version and update the docs.

**Review by pbrisbin — COMMENTED** [review]
Love that this is opt-in. I think that makes it fine that there's a trade-off ("build even on full hit"). I don't follow why this would be a major version bump. It seems to be adding a new option with the same behavior by default... e.g. minor.

**Comment by stackptr** [comment]
> I don't follow why this would be a major version bump. It seems to be adding a new option with the same behavior by default... e.g. minor.

Sounds good; I wasn't sure how to characterize the risk of removing the noop. The behavior is certainly the same and what is changed in the default scenario, not making use of the input, is an implementation detail.

File: action.yml
Comment (stackptr) [reply action.yml:224]: [`success()` is applied by default][0] when an `if` conditional does not use a status check function. [`failure()` is needed][1] in order to run the step after some previous failure, which removes the default `success()` and requires it to be explicit.

Using `outputs.cache-hit` with the status check functions would be:

```suggestion
      if: (success() && steps.restore-deps.outputs.cache-hit != 'true') || (failure() && inputs.cache-save-always)
```

That affects saving the cache when a cache hit occurred and the build succeeded (no longer no-ops). If `inputs.cache-save-always` is used by the action, then it's possible the cache is a partial set of artifacts from a previous failed build. In this case, the successful build has added artifacts that will remain uncached.

[0]: https://docs.github.com/en/actions/learn-github-actions/expressions#status-check-functions
[1]: https://docs.github.com/en/actions/learn-github-actions/expressions#failure-with-conditions

Comment (pbrisbin) [reply action.yml:224]: > when an if conditional does not use a status check function

I didn't know that part. That makes it complicated, but I guess I get it.

> That affects saving the cache when a cache hit occurred and the build succeeded (no longer no-ops). If inputs.cache-save-always is used by the action, then it's possible the cache is a partial set of artifacts from a previous failed build. In this case, the successful build has added artifacts that will remain uncached

I'm having trouble parsing this. You're saying the default behavior (using this action with `cache-save-always: false`) is changed or not? If it is changed, is there a reason it can't be not-changed?

Comment (stackptr) [reply action.yml:224]: The default behavior with `cache-save-always: false` is not changed, but using `outputs.cache-hit` as suggested introduces the aforementioned problem when `cache-save-always: true`. One way to address it would be to use a different `key` when saving the cache after a build failure, versus what is currently used when the build has succeeded, and using this different `key` in `restore-keys`. This would allow restoring the use of `outputs.cache-hit` in both the build and the cache-save steps, but I don't think the added complexity is worthwhile. `stack build` should _essentially_ noop if a complete cache of build artifacts was previously restored.

Comment (pbrisbin) [reply action.yml:224]: > The default behavior with `cache-save-always: false` is not changed

Great, love it. Merge away and minor version bump :hear_no_evil: 

> but

I think what you're saying here is just that `cache-save-always: true` results in us saving the cache even on a primary key hit. Right? If so, great, love it, don't even consider it an issue. The option literally says to save "always", which sounds like what that is.

Comment (pbrisbin) [inline action.yml:224]: Sorry, don't merge away yet.

I think inputs are always strings. So `"false"` will still come out as truthy. I think you want to use `== 'true'` or `!= 'false'` here to get the behavior you want.

Should we add a test case?

**Review by pbrisbin — CHANGES_REQUESTED** [review]

File: action.yml
Comment (stackptr) [reply action.yml:224]: > I think what you're saying here is just that `cache-save-always: true` results in us saving the cache even on a primary key hit. Right?

Not exactly. The semantics of the input is to save the cache _even when the build has failed_, regardless of the primary key hit.

I have some confusion about what is considered "behavior" and therefore what changes to the default behavior necessitate a major version bump. If this PR was released as a minor change, workflows using `@v4` would see exactly four changes when inspecting the logs:

- When the dependencies cache is restored via the primary key, the step to build dependencies runs.
- When the dependencies cache is restored via the primary key, the step to save the artifacts from building dependencies runs.
- When the build cache is restored via the primary key, the step to build the project runs.
- When the build cache is restored via the primary key, the step to save the artifacts from building the project runs.

Previously, each of the steps in the aforementioned scenario would no-op. I _think_ that this is not a meaningful difference in behavior, because the expectation is that the action builds/tests a project. Running `stack build` or saving the cache, instead of a no-op, should have no further impact than the short interval where `stack build` determines there is nothing left to build.

The argument from the other direction and the reason I initially recommend a major version bump, is that going from a no-op to a step that runs _anything_ is a place where bugs could now arise, however unlikely that might be.

My opinion is that a workflow using something other than a fully-specified version or hash opts-in to some possibility that upstream updates result in some bugs, and trust that maintainers try not to introduce those bugs. In this case, a major version bump seems overly defensive.

Comment (stackptr) [reply action.yml:224]: I think you're right. A test case added to `example.yml` is a good call, but testing the input seems like it requires either introducing another example stack project so that a failure occurs, or somehow triggering a failure in the existing project.

Since I would like to adopt this input in `megarepo`, I'll start a branch to use this input and change something in the project to demonstrate the build failing.

Comment (stackptr) [reply action.yml:224]: Testing this input here: https://github.com/freckle/megarepo/pull/32424

Comment (stackptr) [reply action.yml:224]: The [CI logs][0] show the `actions/cache/save` occurring after `stack build` fails.

[0]: https://github.com/freckle/megarepo/actions/runs/7589347043/job/20673802594?pr=32424

Comment (pbrisbin) [reply action.yml:224]: I think I see.

"behavior" here, at least the behavior I care about, are steps that were a no-op before, and no do things. That's a pretty important behavior. Because you've listed 4 steps that would previously no-op and now do things, then when you said this:

> The default behavior with `cache-save-always: false` is not changed

I think that's NOT true.

> The argument from the other direction and the reason I initially recommend a major version bump, is that going from a no-op to a step that runs anything ~is a place where bugs could now arise, however unlikely that might be.~
>
> My opinion is that a workflow using something other than a fully-specified version or hash opts-in to some possibility that upstream updates result in some bugs

I agree, but this is not related to bugs, it's because steps take time. We work pretty hard on this action to make it not do unnecessary work and therefore be as fast as possible. Being able to opt into a configuration where you cache more, but possibly run slower is great, but having the default behavior change such that more work is done and therefore the action takes longer is not.

And this, IMO, is not about major-vs-minor, it's actually about, can we please try to avoid that? I'm not sure why we can't keep the no-op behavior when `cache-save-always: false` we have all the information there, why can't we build the appropriate `if` to get there?

Comment (pbrisbin) [reply action.yml:224]: > The [CI logs](https://github.com/freckle/megarepo/actions/runs/7589347043/job/20673802594?pr=32424) show the actions/cache/save occurring after stack build fails.

Cool, but I think the important thing to test is that it *doesn't* happen when you don't configure save-always. That would be the bug that treating `"false"` a a truthy non-empty string would cause. Is there an example of that you can link to?

Comment (stackptr) [reply action.yml:224]: I think this expression must be erroneous because if these were actually booleans and not strings, the existing `!= 'true'` conditional would always fail...

Comment (pbrisbin) [reply action.yml:224]: > if these were actually booleans and not strings, the existing != 'true' conditional would always fail...

Why do you expect that? I think they are actually strings, not booleans.

Comment (stackptr) [reply action.yml:224]: I'm saying your suspicion about this being incorrect is true, they should indeed be strings.

Comment (pbrisbin) [reply action.yml:224]: Oh OK.

Comment (stackptr) [reply action.yml]: > can we please try to avoid that? I'm not sure why we can't keep the no-op behavior when `cache-save-always: false`; we have all the information there, why can't we build the appropriate `if` to get there?

Sure. Restoring the no-op behavior requires being able to differentiate the cache saved after a build failure versus after a success. I've added a `-partial` suffix to be set on the cache key to achieve this, but this requires a little added complexity in the form of an additional step to set the suffix, since status check functions can only be used in an `if` conditional. These steps are only run when `cache-save-always: true`, so nothing changes in the default behavior.

The workflow in this repo shows the action run without this new input, with or without a full cache-hit, and the subsequent keys used when this suffix is not set. https://github.com/freckle/megarepo/pull/32424 has a CI run with [`cache-save-always: false` and build failure][0] (no cache is saved) and [`cache-save-always: true` and build failure][1] (saving with a `-partial` key).

[0]: https://github.com/freckle/megarepo/actions/runs/7616139302/job/20742234779?pr=32424
[1]: https://github.com/freckle/megarepo/actions/runs/7616139302/job/20742235580

Comment (stackptr) [reply action.yml]: After some conversation over Slack, I've reverted some changes around using different cache keys and have instead extended the `if` conditionals toward some new behavior when the added input is set, which should keep the default behavior without this input the same.

File: action.yml
Comment (pbrisbin) [inline action.yml:215]: I worry this doesn't work (the yaml `|` syntax). Isn't this just equivalent to `if: "some long string"`?

I believe the documentation says that `if: ${{ something }}` can be reduced to `if: something` only if `something` is "trivial" or "simple" or something. This seems very not trivial to me, so I'm shocked if it works.

I'll approve, assuming you've checked the example workflows and seen it's working as expected.

File: README.md
Comment (pbrisbin) [inline README.md:54]: Maybe mention that this will cause additional building on cache-hits (and why). And WDYT about putting a (since <version>) parenthetical?

**Review by pbrisbin — APPROVED** [review]

File: action.yml
Comment (stackptr) [reply action.yml:215]: I think GitHub does some magic to allow this long YAML string to evaluate as a conditional:

When `cache-save-always: false`, a [build failure is observed to bypass saving the cache][0] (see [821cf4d][1], which has the input turned off for `backend-build`). If this was evaluated as a long string, it would always be truthy and so these steps would never no-op. In that same failing build, `cache-save-always` is enabled and the CI run for `backend-test` shows the cache-save step running after the failure. This wouldn't happen for an `if` conditional evaluated as a string since that string would have an implicit `success()` status check function applied.

However, I can't quite verify the default action behavior by removing the build failure and setting `cache-save-always: false`. While the [first CI run for `backend-build`][3] has an expected cache-miss, subsequent builds show the cache key changing, apparently due to the computed `sources-hash` of the cache key. I'm not sure if this has been introduced by this change, but I'll investigate further.

[0]: https://github.com/freckle/megarepo/actions/runs/7618177840/job/20748796709
[1]: https://github.com/freckle/megarepo/pull/32424/commits/821cf4d9c3c4e7c52552cdeb6ccc3dd1290bc0a5
[2]: https://github.com/freckle/megarepo/actions/runs/7618177840/job/20748797130
[3]: https://github.com/freckle/megarepo/actions/runs/7618711349/job/20750418098

File: action.yml
Comment (stackptr) [reply action.yml]: I think my inability to see a full cache-hit occur is unrelated to this PR or the repro producing those earlier CI runs. In this repo, I've looked at `test (ubuntu-latest, lts)`. After a successful initial run on d36a2d5, I looked at the cache key chosen for the main `stack build` (not dependencies):

- [Retry 1][0] uses `key: lts/Linux-stack-build-94486db893527ccfb2eb0bd863d5e67fd2db2a6f1ea17fb29815ed4ed34b9ae9-bac2141ac4b467e165d93f6fb80d35d9ddab24ae16ba43291dff3e289219baa9-817889aa22ef22a7891ffa6f041b5001f4b769f7e747e28c8b339a918e098250`
- [Retry 2][1] uses `key: lts/Linux-stack-build-94486db893527ccfb2eb0bd863d5e67fd2db2a6f1ea17fb29815ed4ed34b9ae9-bac2141ac4b467e165d93f6fb80d35d9ddab24ae16ba43291dff3e289219baa9-a8d203ecdad23a840690f55fc64fe52997690acbfc22aef34cc023058d70a634`
- [Retry 3][2] uses `key: lts/Linux-stack-build-94486db893527ccfb2eb0bd863d5e67fd2db2a6f1ea17fb29815ed4ed34b9ae9-bac2141ac4b467e165d93f6fb80d35d9ddab24ae16ba43291dff3e289219baa9-7cb01ff854833118b39dcd768d6e7bde63da2c3e9b68539977ab778763a5c347`
- [Retry 4][3] uses `key: lts/Linux-stack-build-94486db893527ccfb2eb0bd863d5e67fd2db2a6f1ea17fb29815ed4ed34b9ae9-bac2141ac4b467e165d93f6fb80d35d9ddab24ae16ba43291dff3e289219baa9-0bcad1e2578b1528b1af397a3419c282db865812a77db209e9b189bd5cd2f994`

Each `key` is the same except for the last hash, which is `steps.setup.outputs.sources-hash`. This is computed as:

```bash
echo "sources-hash=${{ hashFiles('**', '!**/.stack-work') }}" >>"$GITHUB_OUTPUT"
```

I have no idea why this would change from run to run, but I am pretty sure this PR is not introducing this behavior and that it should be sorted out separately.

[0]: https://github.com/freckle/stack-action/actions/runs/7619297684/job/20785923748?pr=34
[1]: https://github.com/freckle/stack-action/actions/runs/7619297684/job/20786256605?pr=34
[2]: https://github.com/freckle/stack-action/actions/runs/7619297684/job/20786323717?pr=34
[3]: https://github.com/freckle/stack-action/actions/runs/7619297684/job/20786376797?pr=34

Comment (pbrisbin) [reply action.yml]: > I have no idea why this would change from run to run

Is the `hashFiles` call picking up content in `.git`, which will change with every commit?

Could add another `'!./.git'` pattern there, but I agree it's something we can figure out separately.

File: action.yml
Comment (stackptr) [reply action.yml:215]: Made this comment in the wrong thread: https://github.com/freckle/stack-action/pull/34#discussion_r1463861279

**Pull request merged**
